modules/etcd/etcd.go

```diff
@@ -0,0 +1,266 @@
+		// this prevents the same option from being applied multiple times (e.g. updating the current node)
+		if apply, ok := opt.(Option); ok {
+			etcdOpts = append(etcdOpts, apply)
+		} else {
+			clusterOpts = append(clusterOpts, opt)
+		}
+	}
+
+	if settings.clusterNetwork == nil { // the first time the network is created
+		newNetwork, err := tcnetwork.New(ctx)
+		if err != nil {
+			return clusterOpts, fmt.Errorf("new network: %w", err)
+		}
+
+		// set the network for the first node
+		settings.clusterNetwork = newNetwork
+
+		clusterOpts = append(clusterOpts, withClusterNetwork(newNetwork)) // save the network for the next nodes
+	}
+
+	// we finally need to re-apply all the etcd-specific options
+	clusterOpts = append(clusterOpts, withClusterOptions(etcdOpts))
+
+	return clusterOpts, nil
+}
+
+// configureCMD configures the etcd command line arguments, based on the settings provided,
+// in order to create a cluster or a single-node instance.
+func configureCMD(settings options) []string {
+	cmds := []string{"etcd"}
+
+	scheme := "http"
+
+	if len(settings.Nodes) < 1 {
+		cmds = append(cmds, "--name=default")
+	} else {
+		clusterCmds := []string{
+			"--name=" + settings.Nodes[settings.currentNode],
+			"--initial-advertise-peer-urls=" + scheme + "://" + settings.Nodes[settings.currentNode] + ":" + PeerPort,
+			"--advertise-client-urls=" + scheme + "://" + settings.Nodes[settings.currentNode] + ":" + ClientPort,
+			"--listen-peer-urls=" + scheme + "://0.0.0.0:" + PeerPort,
+			"--listen-client-urls=" + scheme + "://0.0.0.0:" + ClientPort,
+			"--initial-cluster-state=new",
+		}
+
+		clusterStateValues := []string{}
+		for _, node := range settings.Nodes {
+			clusterStateValues = append(clusterStateValues, node+"="+scheme+"://"+node+":"+PeerPort)
+		}
+		clusterCmds = append(clusterCmds, "--initial-cluster="+strings.Join(clusterStateValues, ","))
+
+		if settings.ClusterToken != "" {
+			clusterCmds = append(clusterCmds, "--initial-cluster-token="+settings.ClusterToken)
+		}
+
+		cmds = append(cmds, clusterCmds...)
+	}
+
+	if settings.MountDataDir {
+		cmds = append(cmds, "--data-dir="+DataDir)
+	}
+
+	cmds = append(cmds, settings.AdditionalArgs...)
+
+	return cmds
+}
+
+// ClientEndpoint returns the client endpoint for the etcd container, and an error if any.
+// For a cluster, it returns the client endpoint of the first node.
+func (c *EtcdContainer) ClientEndpoint(ctx context.Context) (string, error) {
+	host, err := c.Host(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	port, err := c.MappedPort(ctx, ClientPort)
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("http://%s:%s", host, port.Port()), nil
+}
+
+// MustClientEndpoint returns the client endpoint for the etcd container.
+// For a cluster, it returns the client endpoint of the first node.
+// This method panics if an error occurs.
+func (c *EtcdContainer) MustClientEndpoint(ctx context.Context) string {
```

Comment:
question: given these could easily fail do we want to expose `Must` versions?

Reply:
I'm not sure if I follow: you mean exposing them could lead to more errors?

Reply:
Must is typically resolved for cases which only happen due coding bug, for example [regexp.MustCompile](https://pkg.go.dev/regexp#MustCompile), you have to give it a bad regexp for it to fail. These can fail due to runtime conditions, which in turn would panic a whole app or test run which isn't desirable.

I would only include the non-must versions to prevent the user shooting themselves in the foot by using them.

Reply:
Understood.
> These can fail due to runtime conditions, which in turn would panic a whole app or test run which isn't desirable.